validator/delphi_validator/validate.py

```diff
@@ -1,237 +1,721 @@
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        switcher = {
+            'unweighted': timedelta(days=1),
+            'weighted': timedelta(days=4)
+        }
+        # Get the setting from switcher dictionary
+        thres = switcher.get(
+            weighted_option, lambda: "Invalid weighting option")
+
+        if max_date < self.generation_date - thres:
+            self.raised_errors.append(ValidationError(
+                ("check_min_max_date", geo, sig),
+                max_date.date(),
+                "most recent date of generated file seems too long ago"))
+
+    def check_max_allowed_max_date(self, max_date, geo, sig):
+        """
+        Check if time since data was generated is reasonable or too recent.
+
+        Arguments:
+            - max_date: date of most recent data to be validated; datetime format.
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        if max_date > self.generation_date - timedelta(days=1):
+            self.raised_errors.append(ValidationError(
+                ("check_max_max_date", geo, sig),
+                max_date.date(),
+                "most recent date of generated file seems too recent"))
+
+    def check_max_date_vs_reference(self, df_to_test, df_to_reference, checking_date, geo, sig):
+        """
+        Check if reference data is more recent than test data.
+
+        Arguments:
+            - df_to_test: pandas dataframe of a single CSV of source data
+            (one day-signal-geo_type combo)
+            - df_to_reference: pandas dataframe of reference data, either from the
+            COVIDcast API or semirecent data
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        if df_to_test["time_value"].max() < df_to_reference["time_value"].max():
+            self.raised_errors.append(ValidationError(
+                ("check_max_date_vs_reference", checking_date.date(), geo, sig),
+                (df_to_test["time_value"].max(),
+                 df_to_reference["time_value"].max()),
+                'reference df has days beyond the max date in the =df_to_test=; ' +
+                'checks are not constructed to handle this case, and this situation ' +
+                'may indicate that something locally is out of date, or, if the local ' +
+                'working files have already been compared against the reference, ' +
+                'that there is a bug somewhere'))
+
+    def check_rapid_change(self, df_to_test, df_to_reference, checking_date, geo, sig):
```

Comment:
Rename to specify that it's a change in num_rows? this is not the only rapid change we might test for.